test/dynamo/test_guard_serialization.py

```diff
@@ -219,6 +383,82 @@ def fn(m, x):
         self._test_check_fn(ref, loaded, {"m": GlobalModule()}, True)
         self._test_check_fn(ref, loaded, {"m": torch.nn.Module()}, False)
 
+    def test_tensor_subclass_metadata_match(self):
+        class LocalSubclass(torch.Tensor):
+            @staticmethod
+            def __new__(cls, a, outer_size=None, outer_stride=None):
+                if outer_size is None:
+                    outer_size = a.size()
+                if outer_stride is None:
+                    outer_stride = a.stride()
+
+                shape = outer_size
+                kwargs = {}
+                kwargs["strides"] = outer_stride
+                kwargs["storage_offset"] = a.storage_offset()
+                kwargs["device"] = a.device
+                kwargs["layout"] = a.layout
+                kwargs["requires_grad"] = a.requires_grad
+                kwargs["dtype"] = a.dtype
+                return torch.Tensor._make_wrapper_subclass(cls, shape, **kwargs)
+
+            def __init__(self, a, outer_size=None, outer_stride=None):
+                self.a = a
+
+            @classmethod
+            def __torch_dispatch__(cls, func, types, args, kwargs):
+                if kwargs is None:
+                    kwargs = {}
+                args_a = pytree.tree_map_only(LocalSubclass, lambda x: x.a, args)
+                kwargs_a = pytree.tree_map_only(LocalSubclass, lambda x: x.a, kwargs)
+                out_a = func(*args_a, **kwargs_a)
+                if isinstance(out_a, torch.Tensor):
+                    return LocalSubclass(out_a)
+                return out_a
+
+            def __tensor_flatten__(self):
+                return ["a"], None
+
+            @staticmethod
+            def __tensor_unflatten__(inner_tensors, meta, outer_size, outer_stride):
+                assert meta is None
+                a = inner_tensors["a"]
+                if type(a) is torch.Tensor:
+                    assert outer_size is not None
+                    assert outer_stride is not None
+                return LocalSubclass(a, outer_size, outer_stride)
+
+        def fn(x):
+            return x * 2
+
+        local_sub = LocalSubclass(torch.randn(3))
+        with self.assertRaisesRegex(
+            RuntimeError, "Please define the class at global scope"
+        ):
+            self._test_serialization("TENSOR_SUBCLASS_METADATA_MATCH", fn, local_sub)
+
+        # use TwoTensor as an example of a subclass with None for extra metadata
+        from torch.testing._internal.two_tensor import TwoTensor
+
+        tt = TwoTensor(torch.randn(3), torch.randn(3))
+        self._test_serialization("TENSOR_SUBCLASS_METADATA_MATCH", fn, tt)
```

Comment:
Regarding unittesting, _test_serialization() only tests whether dynamo can finish a roundtrip of the given type of guard. To verify the reconstructed guards have the same behavior, we can also take the returned guard managers from _test_serialization() and use _test_check_fn() to compare the behavior of original guard manager vs deserialized guard manager (like other unittest do). 
Sorry maybe it's not super obvious to do these in two steps, feel free to refactor the unittest code if you have better idea how to test these.

Reply:
oh whoops that makes sense. Dumb mistake on my part; will fix. I don't think it's -too- unwieldy to have this in 2 steps, but maybe I'll put a little thought into streamlining this